### ISSUE TITLE
Restrict insect data entry to bees and hoverflies

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/PoMS#163]
For the insect data entry form:
https://ukpoms.org.uk/node/44/edit

we currently filter the allowed taxa to all Hymenoptera and all flies, using the UKSI informal taxon groups:

 ![image](https://github.com/BiologicalRecordsCentre/PoMS/assets/16039111/256a38c5-33a7-4c5a-a620-7ced974cf93b)

I would like to restrict this to just bees and hoverflies, to make data entry easier and less prone to errors.

John has previously told me that "The current search API for taxon names allows a search limited to taxon group, a list of taxon external keys, or a scratchpad list. Or, we would need to update the code to allow search by a list of orders."

Questions:
1. If I get a list of all the taxon external keys for bees and hoverfies it will be around 600 keys. Will that work for filtering on the form or is too long a list?
2. Or would it be better to use a scratchpad list? If so, how do we create such a list (or is that something that is only available on iRecord)?

**Comment by BirenRathod** [comment]
@kitenetter & @andrewvanbreda Info: At preset, we are experiencing a similar problem on the BeeWalk site using a filter for just bumblebees where the species filter produces a list of species of around a similar number of external keys. so may be a too-long list. Once sorted it should be fine.

**Comment by andrewvanbreda** [comment]
Hi @kitenetter 

I do not know whether it would be technically too long or a problem speed wise. However 600 does sound quite long from the point of view of practicality in a filter field.
We could try it though! 
It is not hard to try as an experiment as it is very easy to create a multidev clone of live in Pantheon (just a few minutes).

I think if we were to go that route though we would want to use Organism Key anyway rather than external key?

I think a scratchpad is a possibility, I don't think it is iRecord only. It is something I coded for Ireland years ago, but I would need to refresh my memory to determine if it definetely was suitable.

Perhaps the other possibility is to use the same lists that EU PoMS Spring use on that very same equivilent page on the SPRING site?

Lists

https://warehouse1.indicia.org.uk/index.php/taxon_list/edit/272

https://warehouse1.indicia.org.uk/index.php/taxon_list/edit/273

Andy

**Comment by kitenetter** [comment]
On iRecord we already have a scratchpad for bees:
https://irecord.org.uk/scratchpads/edit-scratchpad?scratchpad_list_id=3874

and it would be easy to set one up that added the hoverflies. But I don't know how the scratchpads interact with the rest of Indicia, e.g. whether a scratchpad created in iRecord can be used on the PoMS website.

I thought John had told me that the filters have to use the external key rather than the organism key but I may be misunderstanding.

The EU lists won't work as they include species we don't have in the UK, and also lose the link to UKSI which we need to retain. Presumably we could set up a new taxon list in the warehouse for PoMS use that is a subset of UKSI? But I'm not familiar with how this is done.

If the scratchpad lists can be used that is simplest from my perspective, but otherwise I can try to sort out a list of suitable keys (but I don't currently have access to pgAdmin so that may delay things).

**Comment by andrewvanbreda** [comment]
Hi @kitenetter 
OK I will have to refresh my mind on Scratchpads and let you know.

Organism Keys can be used as filter, we are going to be doing that for NPMS.

We used a sublist in NPMS, but it has caused problems so we are changing to Organism Key.

You can see that option in the list here

<img width="573" alt="Screen Shot 2023-11-30 at 15 11 38" src="https://github.com/BiologicalRecordsCentre/PoMS/assets/2463614/0d59cf5b-e4ce-45ee-9e04-836ede5831de">


I will investigate scratchpads and let you know.

**Comment by andrewvanbreda** [comment]
Hi @kitenetter ,

I have looked up and refreshed my memory on scratchpads.
(I can also see that there is also some new functionality since I worked on it a few years back)

My observations are

1A. There is a general Indicia form for setting up a scratchpad.
This is the same screen you see on iRecord, but it is not just restricted to iRecord I don't think.
It looks to be easy to use as is in the Indicia general forms.
As you can see, you just give it a list of names.

1B. Unfortunately I couldn't get the form to work on my machine. This could be that I am doing something wrong, especially if it is successfully being used on iRecord.
This did restrict me from testing it further though.

2. There seems to be a code extension that can be added to a data entry form that would load the species list from a scratchpad.
So it does not work like a normal filter, but it does look like it does a similar thing in that it loads particular species onto a grid.
(I did not try this though because the scratchpad setup form wasn't working on my machine, but let's assume we can use that).

3. I do have one thought about using the scratchpad though (that would not be a problem with an organism key filter).
Is your species list always going to want show the preferred name for a species?
If we use a scratchpad, it is possible you will be stuck showing the name the scratchpad is pointing at even if it is no longer the preferred name.
I am not quite sure, it is something I would have to try out to be certain.
I would hope it would just respect the type of name we select to show on the species grid (that is an option in the data entry screen configuration options (e.g. Preferred, Default Common etc)) 

So give the above, shall I go ahead and give the scratchpad a proper try?

Perhaps if you give me 10 species you want to include I can use that as a starting point.

Then if that fails we could fall back on external/organism key filtering?

Andy

**Comment by kitenetter** [comment]
@andrewvanbreda I've just set up a scratchpad in iRecord (https://irecord.org.uk/scratchpads/edit-scratchpad?scratchpad_list_id=8380) which uses these family names:
Andrenidae
Apidae
Colletidae
Halictidae
Megachilidae
Melittidae
Syrphidae

I would expect that to include all the species we need, and would hope that it pulls in the current recommended names (although I would expect the scratchpad to need editing if the family-level taxonomy was to change in future).

If you need some species level names to try I can supply some, but I'd hope that it works at family level?

**Comment by andrewvanbreda** [comment]
@kitenetter 
OK great, I will have a look at this when I get the chance.

**Comment by andrewvanbreda** [comment]
Hi @kitenetter ,

I have had a look at this.

Scratchpads are linked to a particular website (there is a website_id on the database row).
However I have been unable to get the scratchpad page working on my UKPoMS test site.

I can also see on the data entry page it is returning no species, again I think because the website_id is wrong.

What I will do is get Biren to switch the website_id on your scratchpad from iRecord to UKPoMS.
It does mean you probably won't be able to make further edits to the list once that is done.
Once I have got it working, we can switch it back to iRecord for you to finish off the list (assuming I am right about all of this)

@BirenRathod  would you be able to run the following statement please (on live warehouse).
Cheers

Thanks

update indicia.scratchpad_lists
set website_id = 136, updated_on=now(), updated_by_id = 5553
where id = 8380;

**Comment by kitenetter** [comment]
@andrewvanbreda I think Biren is now on leave until the new year so this will have to wait.

**Comment by andrewvanbreda** [comment]
@kitenetter 
I will get someone else to help me like John so I can progress this while you are on leave.

**Comment by BirenRathod** [comment]
@andrewvanbreda did you manage somebody to run this query or shall I do it for you?

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod 
Thanks for checking about this.
Jim ran it for me in the end, so there is no need to do anything.
Thanks again.
Andy

**Comment by andrewvanbreda** [comment]
Hi @kitenetter 
I have looked at this.
I think actually the existing scratchpad code does not do what we want.
It doesn't act as a filter for freetext entry, but instead preloads a species list with the scratchpad rows.
This would only be ok for a small number of species.
You can see this behaviour here,
https://avb-scratch-ceh-ukpoms-d10.pantheonsite.io/species-data-entry

I tried searching Indicia's code and cannot see any code where it would act as a filter.
The code also even displays the following words on the screen 
"The list of species has been loaded into the form for you"
which indicates it is only for loading a static list.

I could perhaps code something new to filter per scratchpad, but I think it might be worth checking if we could use the organism key as a filter first (even if there are a lot of them).
Is it easy for you to get the organism keys for the species you need, or is that going to be a pain?
Or perhaps we could try a quarter of them or something to see how the speed holds up.

**Comment by kitenetter** [comment]
@robin-hutchinson can you export a list from the Indicia version of the UKSI that gives organism keys for all species in these families (bees and hoverflies):
Andrenidae
Apidae
Colletidae
Halictidae
Megachilidae
Melittidae
Syrphidae

**Comment by robin-hutchinson** [comment]
No problem - here it is: [Bee and Hoverfly Organism Keys.csv](https://github.com/BiologicalRecordsCentre/PoMS/files/13840881/Bee.and.Hoverfly.Organism.Keys.csv)

**Comment by andrewvanbreda** [comment]
Thanks @robin-hutchinson, that is really great.

@kitenetter : This is not working, but not in the way I expected.
It seems to be absolutely fine as long as the list isn't longer than 314 rows.
If it is longer than that, then it never returns any species for the user to see.

This is worth me investigating as apart from that, this looks very promising.
The 800+ keys does not seem to be as unmanageable in practice as I thought. Also there doesn't appear to be any performance problems from the list when it is working (314 long).
Also the list does seem to save into the configuration field ok, I had feared the field wouldn't be long enough.
Will get back to you.

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod 
I don't have my connection to the Warehouse since the infrastructure changes. Would you be able to get me the result of the query in the attached file please.
Thanks

[Get taxon meaning IDs.txt](https://github.com/BiologicalRecordsCentre/PoMS/files/13918571/Get.taxon.meaning.IDs.txt)

**Comment by BirenRathod** [comment]
@andrewvanbreda  Here it is. 
[cache_taxa_taxon_lists_202401121247.zip](https://github.com/BiologicalRecordsCentre/PoMS/files/13919306/cache_taxa_taxon_lists_202401121247.zip)

**Comment by andrewvanbreda** [comment]
@BirenRathod Brilliant that is really great. Thanks

**Comment by andrewvanbreda** [comment]
Hi @kitenetter,

I have looked into this.
The problem is the request to the Warehouse server is too long, hence it breaks when there are too many organism keys.
(I think it is possible to allow longer requests, but I think it would mean messing with Warehouse 1's configuration, so I don't think that is going to be the way forward).
As organism keys have quite a few characters, I thought one way of fixing this situation would be to filter based on taxon meaning ID which are much shorter numbers.
This does improve the situation, and it now allows about 550(ish) rows before it starts going wrong.
This leaves me wondering how to get around this problem.
There is a similar page on SPRING (EU Pollinators), in that case the Bee/Hoverfly entries are on different grids.
I am wondering if we could perhaps split this so there are two entry grids, one for bees, one for hoverflies.

I have experimented on that on that test site I created, and indeed a separate filter can be applied to each grid.
So I think this would likely work providing there aren't more than about 550 bees in your list, or 550 hoverflies.

If you agree this might be a way forward, would you be able to send me the list of organism keys for bees/hoverflies but split into two lists.

Thanks

Andy

**Comment by kitenetter** [comment]
Thanks @andrewvanbreda 

In the long term I think it would be best to work on a potential solution that John put forward: "a new feature to allow a higher taxon filter using the cache_taxon_paths table" in [this iRecord issue](https://github.com/BiologicalRecordsCentre/iRecord/issues/1341).

However, until we are able to prioritise that work I agree that having a split into a bee grid and a hoverfly grid should be fine.

@robin-hutchinson please can you re-run your export of organism keys but showing which are bees and which hoverflies, or as two separate files.

**Comment by robin-hutchinson** [comment]
Here it is: [Bee and Hoverfly Keys and IDs.csv](https://github.com/BiologicalRecordsCentre/PoMS/files/13921287/Bee.and.Hoverfly.Keys.and.IDs.csv)

**Comment by andrewvanbreda** [comment]
Hi @robin-hutchinson,
Thanks for doing the list, really useful.
Andy

**Comment by andrewvanbreda** [comment]
Hi @kitenetter ,

I have been testing the page I setup and have come up with a couple of thoughts.

1.
As the species list this page is now running are more limited, there is a potential problem where a user could edit an existing sample, and the occurrences species to load onto the species grid is no longer available.
Do you think this is a problem in actual practice?
If so, I will need to think of a solution (my usual solution to this issue on other projects may not work because of the way UKPoMS works)

2. I have tried to do some tests to make sure the species in the lists are selectable to the user.
I obviously can't test every species, but I have done things like try to enter species at end end of the bee/hoverfly lists you gave me.
However I have noticed a bit of an oddity.

When I reduced the filter on the bees and hoverflies grids to only be a couple of species at each I noticed I could not enter some species I expected to.
However they did work when I then went back and changed it back to use the full filter list you gave me.

I believe this is because some species on the list you gave me don't have the "Allow Data Entry" flag switched on.
I will give an example.
Black Mud Bee
Near the bottom of your list this has a row indicated by
Taxon Meaning ID 270935,Organism Key NBNORG0000100942
In the Warehouse the taxon with this meaning/organism key actually has the Allow Data Entry flag off and the Organism Deprecated flag on, so this item does not appear in the list.
Black Mud Bee does still display but only because it appears on this page in the warehouse
https://warehouse1.indicia.org.uk/index.php/taxa_taxon_list/edit/407722
and that uses taxon meaning ID 177108, organism key NBNORG0000117553 which is also appearing on your list as a duplicate.

How much of a problem do you think this is?
Is there a danger some items in your list not having the Allow Data entry flag on and there not being a duplicate to cover for it like there is for Black Mud Bee.
Do I need to get Biren to query the database to find out how many items do not have the Allow Data Entry flag on in your list or do you think this is all ok?
(Obviously I am limited in the tests I can do but I know this case is not unique.
I cannot tell how much of the list you have given me this affects as I don't have access to the Warehouse1 database at the moment since the server infrastructure changes, so I am doing all this using the Warehouse user Interface).

**Comment by robin-hutchinson** [comment]
[Bee and Hoverfly Keys and IDs - Data Entry.csv](https://github.com/BiologicalRecordsCentre/PoMS/files/13938424/Bee.and.Hoverfly.Keys.and.IDs.-.Data.Entry.csv)

I've just added the data entry flag here!

**Comment by kitenetter** [comment]
@andrewvanbreda I don't think we need to worry about your point 1 - it's possible that there may be some records for previous years that are not in the list but I don't anticipate that there will be further editing of these.

And hopefully Robin's new list will address point 2. We don't need to allow people to record anything other than the current recommended name for each organism. And we don't actually need vernacular names at all, although it is not a problem if they are included.

**Comment by andrewvanbreda** [comment]
Hi @kitenetter @robin-hutchinson 
Thank you for the list and responses.
I believe the page on my test site is now operating correctly. I can add it to the live site if ok with you.
https://avb-scratch-ceh-ukpoms-d10.pantheonsite.io/species-data-entry

**Comment by kitenetter** [comment]
Thanks Andy. That looks good on the test site. There are a couple of checks we need to do before going live.

@robin-hutchinson please can you download the 2023 insect data and check if anyone has entered anything that isn't a bee or a hoverfly.

And I will contact the taxonomists to warn them that we intend to change the page layout.

@andrewvanbreda we should be able to go ahead tomorrow but I will confirm.

**Comment by robin-hutchinson** [comment]
I've just run the check and there is one record that isn't a bee or hoverfly. Occurrence 34401946 is Chalarus argenteus (Diptera: Pipunculidae).

**Comment by kitenetter** [comment]
@andrewvanbreda I'm awaiting information from the person who added the non-hoverfly so that we can correct it, after which we can implement the new forms - will get back to you.

**Comment by kitenetter** [comment]
Non-hoverfly record now corrected. Please implement the new form when you can.

**Comment by andrewvanbreda** [comment]
Hi @kitenetter 
This is live now.
Close if happy.

**Comment by kitenetter** [comment]
Thanks @andrewvanbreda 

It looks like the previously entered data is not visible on the new pages - I was expecting that all previously-entered bees and hoverflies would be available. Is there a reason why this isn't the case?

**Comment by andrewvanbreda** [comment]
Hi @kitenetter 
I have reinstated the old code, I will investigate.

**Comment by andrewvanbreda** [comment]
Hi @kitenetter 
Fixed.
This was because we are now using two grids, but those grids were setup to only show records that had been entered on that grid (which is the usual thing to do)
This did mean records from the old grid wouldn't display though.
I have changed this so the grids attempt to show any record providing it meets the filter requirement (because the grids use different sets of species, the same record won't ever show on both grids at same time).
